pkg/tagger/tagstore.go

```diff
@@ -131,31 +138,65 @@ func (e *entityTags) get(highCard bool) ([]string, []string) {
 	}
 
 	// Cache miss
-	var arrays [][]string
 	var sources []string
-	lowCardCount := 0
+	var tagSortList []tagSort
 
 	for source, tags := range e.lowCardTags {
-		arrays = append(arrays, tags)
-		lowCardCount += len(tags)
 		sources = append(sources, source)
+		for _, t := range tags {
+			tagSortList = append(tagSortList, tagSort{
+				tag:         t,
+				tagName:     strings.Split(t, ":")[0],
+				source:      source,
+				cardinality: false,
+			})
+		}
+	}
+
+	for source, tags := range e.highCardTags {
+		for _, t := range tags {
+			tagSortList = append(tagSortList, tagSort{
+				tag:         t,
+				tagName:     strings.Split(t, ":")[0],
+				source:      source,
+				cardinality: true,
+			})
+		}
 	}
-	for _, tags := range e.highCardTags {
-		arrays = append(arrays, tags)
+
+	lowCardTags := []string{}
+	highCardTags := []string{}
+	for i := 0; i < len(tagSortList); i++ {
+		insert := true
+		for j := 0; j < len(tagSortList); j++ {
+			// if we find a duplicate tag with higher priority we do not insert the tag
+			if i != j && tagSortList[i].tagName == tagSortList[j].tagName &&
+				collectors.CollectorPriorities[tagSortList[i].source] < collectors.CollectorPriorities[tagSortList[j].source] {
+				insert = false
+			}
+		}
+		if insert {
+			if tagSortList[i].cardinality {
+				highCardTags = append(highCardTags, tagSortList[i].tag)
+			} else {
+				lowCardTags = append(lowCardTags, tagSortList[i].tag)
+			}
+		}
 	}
-	tags := utils.ConcatenateTags(arrays)
+
+	tags := append(lowCardTags, highCardTags...)
 
 	// Write cache
 	e.RUnlock()
 	e.Lock()
 	e.cacheValid = true
 	e.cachedSource = sources
 	e.cachedAll = tags
-	e.cachedLow = e.cachedAll[:lowCardCount]
```

Comment:
nitpick, but this allows the `cachedLow` to be a slice of the same backing array `cachedAll` uses, and reduces the memory usage of the tagstore cache. You could keep it easily by setting `e.cachedLow = e.cachedAll[len(lowCardTags)]`, then `lowCardTags` will be GCed